src/undo.js

```diff
@@ -0,0 +1,73 @@
+import {
+    checkForAlpine,
+    componentData,
+    updateOnMutation,
+} from './utils'
+import { DeepDiff } from 'deep-diff'
+
+const AlpineUndoMagicMethod = {
+    start() {
+        checkForAlpine()
+        Alpine.addMagicProperty('track', function ($el) {
+            return function (_propertiesToWatch = []) {
+                // TODO: Possibly scope to specific properties
+                $el.changes = []
+                $el.initialComponentState = componentData($el)
+                $el.previousComponentState = JSON.stringify({})
+
+                updateOnMutation($el, () => {
+                    if ($el.dontTrackForOneTick) {
+                        $el.dontTrackForOneTick = null
+                        return
+                    }
+
+                    const previous = JSON.parse($el.previousComponentState)
+                    const fresh = JSON.parse(JSON.stringify($el.__x.getUnobservedData()))
+                    const changes = DeepDiff.diff(previous, fresh, true)
+                    if (changes) {
+                        $el.changes.push(changes)
+                        $el.previousComponentState = JSON.stringify(fresh)
+                        $el.__x.$data.$nextTick(() => {
+                            $el.dontTrackForOneTick = true
+                            $el.__x.updateElements($el)
+                        })
```

Comment:
Did you try it? I couldn't get the UI to update without it. I tried to comment out each updateElements call inidividually but then it would break. Was thinking maybe a race condition with how the deep-diff applies the patch. Could possibly instead proxy `$el.previousComponentState` or something, then updateElements in the set trap. Will mess around more after writing some tests.

Reply:
Welp, I removed everything now and it works. May have been needed in an early iteration or something. Seems ok without it though.

Reply:
Testing some more now, with them removed the component isn't updated when `$el.changes` is empty, so the undo button will still display. Will look into  another approach though.

Reply:
I see, do we wan't to make $el.changes reactive though? 
It's not part of x-data so, technically speaking, it should neither be reactive nor used on the UI.

Reply:
I think there should be a way for users to know whether an "undo" operation is possible. For example, you might disable or hide an undo button when there are no available changes. Same goes for redo (if that gets implemented).

The current version in this PR works though with that undo functionality. I just had to restructure a few things.

Reply:
We could have $hasChanges which compares the current $el.__x.unosservedData with safeEval($el.getAttribute('x-data')), but maybe it's an overComplication.

It's just that, if you have a computed property counting the number of refreshes, I suspect it will be increased by 2 every time

```javascript
function data() {
    let counter = 0
    return {
       foo: 'bar',
       init() {this.$track()}
       getRefreshes() { return ++counter()}
    }
}
```